src/pypi_mapping/custom_pypi_mapping.rs

```diff
@@ -0,0 +1,143 @@
+use rattler_conda_types::{Channel, PackageUrl, RepoDataRecord};
+use reqwest_middleware::ClientWithMiddleware;
+use std::{collections::HashMap, sync::Arc};
+
+use async_once_cell::OnceCell;
+
+use crate::pypi_mapping::MappingLocation;
+
+use super::{prefix_pypi_name_mapping, MappingMap, Reporter};
+
+pub async fn fetch_custom_mapping(
+    client: &ClientWithMiddleware,
+    mapping_url: &MappingMap,
+) -> miette::Result<&'static HashMap<String, HashMap<String, String>>> {
+    static MAPPING: OnceCell<HashMap<String, HashMap<String, String>>> = OnceCell::new();
+    MAPPING
+        .get_or_try_init(async {
+            let mut mapping_url_to_name: HashMap<String, HashMap<String, String>> =
+                Default::default();
+
+            for (name, url) in mapping_url.iter() {
+                // Fetch the mapping from the server or from the local
+
+                match url {
+                    MappingLocation::Url(url) => {
+                        let response = client
+                            .get(url.clone())
+                            .send()
+                            .await
+                            .into_diagnostic()
+                            .context("failed to download pypi mapping from custom location")?;
+
+                        let mapping_by_name: HashMap<String, String> = response
+                            .json()
+                            .await
+                            .into_diagnostic()
+                            .context("failed to parse pypi name mapping")?;
+
+                        mapping_url_to_name.insert(name.to_string(), mapping_by_name);
+                    }
+                    MappingLocation::Path(path) => {
+                        let contents = std::fs::read_to_string(path)
+                            .into_diagnostic()
+                            .context(format!("mapping on {path:?} could not be loaded"))?;
+                        let data: HashMap<String, String> = serde_json::from_str(&contents)
+                            .unwrap_or_else(|_| {
+                                panic!("Failed to parse JSON mapping located at {path:?}")
+                            });
+
+                        mapping_url_to_name.insert(name.to_string(), data);
+                    }
+                }
+            }
+
+            Ok(mapping_url_to_name)
+        })
+        .await
+}
+
+/// Amend the records with pypi purls if they are not present yet.
+pub async fn amend_pypi_purls(
+    client: &ClientWithMiddleware,
+    mapping_url: &MappingMap,
+    default_conda_forge_channel: Channel,
+    conda_packages: &mut [RepoDataRecord],
+    reporter: Option<Arc<dyn Reporter>>,
+) -> miette::Result<()> {
+    let conda_forge_name = default_conda_forge_channel.canonical_name();
+
+    let conda_forge_packages: Vec<RepoDataRecord> = conda_packages
+        .iter()
+        .filter(|package| package.channel.contains(&conda_forge_name))
+        .cloned()
+        .collect();
```

Comment:
This is not needed anymore right?

Reply:
yes! I've changed it a little so we filter now to get the packages that are not present in the mapping ( to not have useless hit's to prefix.dev mapping ) . Let me know if it's ok